### ISSUE TITLE
Programmer Dvorak or layouts that modify shift/capslock behavior?

### ISSUE THREAD

**Issue by moparisthebest** [tmk/tmk_keyboard#163]
I've been using a [slightly customized tmk_keyboard](https://github.com/moparisthebest/tmk_keyboard/tree/master/ps2_usb) ps/2->usb adapter for regular dvorak for years, it's excellent.

Recently, I've been trying to switch to a [slightly customized version](https://www.moparisthebest.com/kbs/programmer-dvorak-QwertySecondary-NumpadPhone-NoSwap-StandardNums-SwapAt-SwapPipe.svg) of [programmer dvorak](http://www.kaufmann.no/roland/dvorak/), which changes shift/capslock behavior in a number of ways.  It basically requires having a seperate layout for when shift or capslock is activated and not, and supporting each key in both layouts to be sent with or without shift.  It also has capslock work for all keys including keys that it normally doesn't work for, the number keys for instance.

For example:
  1. The 1 key on standard QWERTY layout, when pressed:
    1. (without shift OR capslock) OR (with shift AND capslock), sends Shift+7 for the & character
    2. (with shift OR capslock), sends 1 without shift for the 1 character
  2. The GRV key on standard QWERTY layout, when pressed:
    1. (without shift OR capslock) OR (with shift AND capslock), sends Shift+GRV for the ~ character
    2. (with shift OR capslock), sends Shift+4 for the $ character

I've modified my [uinput-mapper code](https://github.com/moparisthebest/uinput-mapper/blob/master/keymaps/dvorak.py#L50) to handle this at the slightly-higher-than-hardware-but-lower-than-user layer in Linux, but I'd like to handle it in tmk_keyboard as well, does it support something like this out of the box in a way I'm just missing? Or how would you imagine defining/handling layouts like this?

Thanks much!

**Comment by tmk** [comment]
I'm not sure I understand programmer dvorak layout correctly but I think there is no easy way to define a layout like that in tmk keymap system.

Seems the layout require complicated layer transitions and many FN definitions but only 32 FN keys are available currently. Without writing C code in action_function() it is not possible at least. Generally speaking, it is very difficult to define this kind of layout on keyboard hardware, because you need to use multiple keycodes to register a character like '&'. This can became problematic under some situations like; when you want to type '&a' quickly you'll get '&A' with simple naive implementation. I think OS configuration or application is way better solution for complecated layout like that(and macros).

But I'm kindof interested in implementing the layout or helping you at least, if you really need this on tmk firmware.


And I think you must check HaaTa's new firmware with powerful keymapping system. Perhaps, you may be able to define the layout easily.
https://github.com/kiibohd/controller
https://www.overleaf.com/read/zzqbdwqjfwwf

**Comment by haata** [comment]
hasuTMK linked me, so I might as well comment :P

Most definitely you would be able to do this with the KLL.
Theoretically there can be up to uint32 layers, though you'd likely run out of flash before then...

I have an open feature request to use the LED state (CapsLock,ScrollLock,NumLock, etc.) as a trigger to macros / function layers. Feel free to ping me if you don't think I'm working on it fast enough :P

**Comment by moparisthebest** [comment]
I appreciate the feedback and the pointer to action_function(), looking at existing uses of this, I think it can probably be used to implement this *fairly* easily.  The trick will be finding a decent way of defining these types of layouts (figuring out how many layers are needed, and then manually defining them is not something I really want to do...) , so I might need some help with C pre-processor magic tricks to do this in a good way. :)

Basically right now, for every keycode there is simply another keycode to send, it's that simple.  With the new system, there needs to be 4 peices of information for every keycode:

1. Keycode to send when shift is not down (int)
2. Whether to invert shift when sending #1 (boolean)
3. Keycode to send when shift is down (int)
4. Whether to invert shift when sending #3 (boolean) (and this is actually inverted the other way, shift is already down, so you might have to send shift_up, or nothing)

All of this is further (and significantly) complicated by caps_lock, which by default modifies some keys but not all, so all of the keys it does not modify (numbers for instance) need defined as well, because we may or may not have to send shift up or down based on this. (Thinking about this again, if the caps_lock light isn't required, it might be easier in the firmware to never send caps-lock, and instead just constantly hold down shift while it's enabled)

For reference, how I defined them in python was originally taken directly from tmk_keyboard, but for the shift-modifying behavior I modified it [like this](https://github.com/moparisthebest/uinput-mapper/blob/master/keymaps/dvorak.py#L50), parsed it into a decent program-useable format [like this](https://github.com/moparisthebest/uinput-mapper/blob/master/uinputmapper/keymapper.py#L100), which would ideally be implemented in the C pre-processor here.  Also the code that needs ported to action_function() is [here](https://github.com/moparisthebest/uinput-mapper/blob/master/input-read#L147).

kiibohd also looks like it might be fairly easy to implement this type of thing as well, I'll definitly keep an eye on it, but my first shot will probably be with tmk_keyboard as I already have the PS/2 -> USB hardware wired up and ready to go. :)
